src/login/connect_server.h

```diff
@@ -0,0 +1,2257 @@
+    MOG_WARDROBE_7   = 0x0040,
+    MOG_WARDROBE_8   = 0x0080,
+    UNUSED_FEATURE_2 = 0x0100,
+    UNUSED_FEATURE_3 = 0x0200,
+    UNUSED_FEATURE_4 = 0x0400,
+    UNUSED_FEATURE_5 = 0x0800,
+    UNUSED_FEATURE_6 = 0x1000,
+    UNUSED_FEATURE_7 = 0x2000,
+    UNUSED_FEATURE_8 = 0x4000,
+    UNUSED_FEATURE_9 = 0x8000,
+};
+
+// Metadata about each session
+struct session_t
+{
+    // Auth session is not yet needed to be stored
+    // std::shared_ptr<handler_session> auth_session;
+    std::shared_ptr<handler_session> data_session;
+    std::shared_ptr<handler_session> view_session;
+    std::shared_ptr<handler_session> pol_session;
+
+    uint32      accountID                 = 0;
+    uint32      serverIP                  = 0;
+    uint32      requestedCharacterID      = 0;
+    std::string requestedNewCharacterName = "";
+    bool        justCreatedNewChar        = false;
+    bool        versionMismatch           = false;
+
+    std::chrono::time_point<std::chrono::system_clock> authorizedTime = server_clock::now();
+};
+
+// NOTE: This collection of flags is 64-bits wide!
+enum AUTH_COMPONENTS
+{
+    // Information to be sent during login sequence.
+    // Each item is added to the data sequentially,
+    // after username and password, which are mandatory:
+    // [username][password][email]
+    // If you opted not to use email:
+    // [username][password]
+    SEND_EMAIL       = 1 << 0,
+    SEND_HOSTNAME    = 1 << 1,
+    SEND_MAC_ADDRESS = 1 << 2,
+
+    // These flags enable additional workflows between connect and xiloader
+    ENABLE_ACCOUNT_CREATE  = 1 << 3, // Ability for a user to create an account
+    ENABLE_ACCOUNT_DELETE  = 1 << 4, // Ability for a user to delete their account
+    ENABLE_PASSWORD_CHANGE = 1 << 5, // Ability for a user to change their password through xiloader
+    ENABLE_PASSWORD_RESET  = 1 << 6, // Ability for a user to flag their account for a password reset through connect server
+};
+
+// [ip_addr][session_hash] = session
+std::unordered_map<std::string, std::map<std::string, session_t>> authenticatedSessions_;
+
+session_t& get_authenticated_session(std::string ipAddr, std::string sessionHash)
+{
+    return authenticatedSessions_[ipAddr][sessionHash]; // NOTE: Will construct if doesn't exist
+}
+
+// hostname/ip conversion functions
+std::string ip2str(uint32 ip)
+{
+    uint32 reversed_ip = htonl(ip);
+    char   address[INET_ADDRSTRLEN];
+    inet_ntop(AF_INET, &reversed_ip, address, INET_ADDRSTRLEN);
+    return fmt::format("{}", address);
+}
+
+uint32 str2ip(const char* ip_str)
+{
+    uint32 ip;
+    inet_pton(AF_INET, ip_str, &ip);
+
+    return ntohl(ip);
+}
+
+void generateErrorMessage(char* packet, uint16 errorCode)
+{
+    std::memset(packet, 0, 0x24);
+
+    packet[0] = 0x24; // size
+
+    packet[4] = 0x49; // I
+    packet[5] = 0x58; // X
+    packet[6] = 0x46; // F
+    packet[7] = 0x46; // F
+
+    packet[8] = 0x04; // result
+
+    packet[28] = 0x10; // Unknown
```

Comment:
Do we have any idea what this is yet? Worth asking atom0s?

Reply:
Hmm. I suppose it might be worth asking about it. This is just "what was there" in the previous iteration.

Reply:
https://github.com/atom0s/XiPackets/blob/main/lobby/S2C_0x0004_ResponseError.md -> `// This field is never referenced within the client. It was always observed to be 0x10, but the actual value or its purpose is unknown.`
Turns out my comment was apparently correct. I've added a comment and a reference.